File: hackerrank/week_of_code_32/fight_the_monsters.go
package main

import (
	"fmt"
	"sort"
)


func main() {
	var monsterCount, hitDamage, seconds, killedMonsters int
	fmt.Scanf("%d %d %d", &monsterCount, &hitDamage, &seconds)
	monsters, _ := intScanln(monsterCount)
	sort.Ints(monsters)
	//fmt.Println(monsters)

	for _, monster := range monsters {
		// figure out the amount of hits we need
		neededHits := monster / hitDamage

		if monster % hitDamage != 0 {
			// we need one more hit
			neededHits++
		}

		seconds -= neededHits
		if seconds >= 0 {
			// this was a valid kill and we have seconds to go
			killedMonsters++
		} else {
			break
		}
	}

	fmt.Println(killedMonsters)

}

/*
	Read an array of integers
 */
func intScanln(n int) ([]int, error) {
    x := make([]int, n)
    y := make([]interface{}, len(x))

	// Make everything in y a reference to x
    for i := range x {
        y[i] = &x[i]
    }
	// Fill y
    n, err := fmt.Scanln(y...)
    x = x[:n]
    return x, err
}

File: hackerrank/week_of_code_32/fight_the_monsters.py
"""
7 8 6
16 19 7 11 23 8 16
"""
_, hit_dmg, seconds = [int(p) for p in input().split()]
monsters = [int(p) for p in input().split()]

monsters_killed = 0
for monst in sorted(monsters):
    needed_hits = 0
    if monst % hit_dmg != 0:
        needed_hits = (monst // hit_dmg) + 1
    else:
        needed_hits = monst // hit_dmg
    seconds -= needed_hits
    if seconds >= 0:
        monsters_killed += 1
    else:
        break

print(monsters_killed)
